### PR TITLE
Smarter lowercase checks and storage for users, communities, and channels

### LINKED ISSUE #2196: usernames are capitalization-specific
if you go to spectrum.chat/users/traykov it says "no user", but if you go to spectrum.chat/users/Traykov it works.

### PR DESCRIPTION
**Pull request by brianlovin**
**Status**
- [ ] WIP
- [x] Ready for review
- [ ] Needs testing

**Deploy after merge (delete what needn't be deployed)**
- api
- hyperion (frontend)

**Run database migrations (delete if no migration was added)**
YES

**Release notes for users (delete if codebase-only change)**
- Fixes a bug where the capitalization in urls would change the results returned from the server

This needs some extra brainpower - here's the situation:
- we aren't validating case when fetching communities, channels, or users
- we aren't validating case when creating communities, channels, or users

While we do have some client-side handlers, we aren't actually ensuring things are lowercase on the server.

This PR does the following:
- Runs a one-time migration to lowercase all usernames and slugs in the db
- Handles `toLowerCase()` at the `models` layer
- Because of the way `dataLoader`s match results to inputs, we *also* have to add `toLowerCase` in some of the root queries (e.g. `rootUser.js`) so that all inputs are *also* lowercase.

This is also another reason we should start to move towards shared db queries soon so that this logic doesn't get mismatched in other places.

I'd love feedback on:
- Does this fix the underlying problem? Or are there better abstractions for handling scenarios like this?
- Am I missing cases where case matters in querying data on the frontend?
- Is my approach of ensuring lowercase appropriate at the `model` layer? Or should it be somewhere else?
- Any better approaches for handling `dataLoader`s to make sure that all inputs are lowercase as well? 

Should fix #2196

### CODE REVIEW

**Comment by brianlovin** [comment]
To demonstrate the problems that this has brought: we have 25 users with the same username.

```js
r.db('spectrum')
  .table('users')
  .group('username')
  .ungroup()
  .filter(row => row('reduction').count().gt(1))
  .pluck('group')
```

I will go through and resolve these manually after this ships and the migrations are run, basically deferring to the earlier user getting the username and following up with people who may lose their username.

**Review by mxstbr — DISMISSED** [review]
We dummies. Overall, this looks like a good approach to me—we should introduce a custom GraphQL scalar called "LowercaseString" which will automatically lowercase any input sent to it automatically so we don't have to do it manually in the resolvers.

**Comment by brianlovin** [comment]
@mxstbr those last two commits *should* do the trick - I clicked all around the app and tried swapping in slugs/usernames with mixed uppercase/lowercase and everything resolved just fine. Mind giving this a quick pull and making sure things are good on your end?

File: api/models/community.js [hunk 1]
```diff
@@ -321,7 +321,7 @@ export const editCommunity = ({
     .then(result => {
       return Object.assign({}, result, {
         name,
-        slug,
+        slug: slug,
```
Comment (mxstbr) [inline api/models/community.js]: nit: unnecessary

File: api/models/community.js [hunk 2]
```diff
@@ -155,7 +155,7 @@ export const createCommunity = (
         website,
         profilePhoto: null,
         coverPhoto: null,
-        slug,
+        slug: slug,
```
Comment (mxstbr) [inline api/models/community.js]: nit: unnecessary

**Review by mxstbr — DISMISSED** [review]
Yep this is great! :tada:

**Comment by brianlovin** [comment]
Wow this is the first time in a long time i've seen all e2e tests pass! I'll fix those nits and admin ship.

**Comment by brianlovin** [comment]
I have in my notes that to run the migrations on this *after* shipping to prod. Will put on alpha in the meantime.

**Comment by brianlovin** [comment]
Migration run, working through duplicate usernames now :(

**Pull request merged**